app/models/concerns/taggable.rb

```diff
@@ -86,12 +86,20 @@ def process_tags(tag_list)
     return unless tag_list.present?
 
     Rails.logger.info "Processing tags: #{tag_list} for record: #{id}"
-    tags = tag_list.compact_blank.join(",")
-
-    raise ArgumentError, "Invalid tags" unless valid_tags?(tags)
-    set_tag_list_on(language.code.to_sym, tags)
+    removed_tags = current_tags_list - tag_list
+    cognates_hash = cognates_names_for(tag_list + removed_tags)
+    tag_list_without_redundant_cognates = tag_list - cognates_hash.slice(*removed_tags).values.flatten
+    tag_list_with_cognates_to_add = tag_list_without_redundant_cognates + cognates_hash.slice(*tag_list_without_redundant_cognates).values.flatten
+    final_tag_list = tag_list_with_cognates_to_add.uniq.compact_blank.join(",")
+    set_tag_list_on(language.code.to_sym, final_tag_list)
     save!
   end
 
-  def valid_tags?(tags) = true
+  def cognates_names_for(tag_name_list)
+    hash = {}
+    Tag.where(name: tag_name_list).each do |tag|
```

Comment:
We can make it even more idiomatic)
```suggestion
    Tag.where(name: tag_name_list).each_with_object({}) do |hash, tag|
      hash[tag.name] = tag.cognates_tags.for_context(language.code.to_sym).uniq.pluck(:name).push(tag.name)
    end
```